protocols/request-response/src/handler.rs

```diff
@@ -302,41 +387,34 @@ where
         }
 
         // Check for inbound requests.
-        while let Poll::Ready(Some(result)) = self.inbound.poll_next_unpin(cx) {
-            match result {
-                Ok(((id, rq), rs_sender)) => {
-                    // We received an inbound request.
-                    self.keep_alive = KeepAlive::Yes;
-                    return Poll::Ready(ConnectionHandlerEvent::NotifyBehaviour(Event::Request {
-                        request_id: id,
-                        request: rq,
-                        sender: rs_sender,
-                    }));
-                }
-                Err(oneshot::Canceled) => {
-                    // The inbound upgrade has errored or timed out reading
-                    // or waiting for the request. The handler is informed
-                    // via `on_connection_event` call with `ConnectionEvent::ListenUpgradeError`.
-                }
-            }
+        if let Poll::Ready(Some((id, rq, rs_sender))) = self.inbound_receiver.poll_next_unpin(cx) {
+            // We received an inbound request.
+            self.keep_alive = KeepAlive::Yes;
+            return Poll::Ready(ConnectionHandlerEvent::NotifyBehaviour(Event::Request {
+                request_id: id,
+                request: rq,
+                sender: rs_sender,
+            }));
         }
 
         // Emit outbound requests.
-        if let Some(request) = self.outbound.pop_front() {
-            let info = request.request_id;
+        if let Some(request) = self.pending_outbound.pop_front() {
+            let protocols = request.protocols.clone();
+            self.requested_outbound.push_back(request);
+
             return Poll::Ready(ConnectionHandlerEvent::OutboundSubstreamRequest {
-                protocol: SubstreamProtocol::new(request, info)
-                    .with_timeout(self.substream_timeout),
+                protocol: SubstreamProtocol::new(Protocol { protocols }, ()),
             });
         }
 
-        debug_assert!(self.outbound.is_empty());
+        debug_assert!(self.pending_outbound.is_empty());
 
-        if self.outbound.capacity() > EMPTY_QUEUE_SHRINK_THRESHOLD {
-            self.outbound.shrink_to_fit();
+        if self.pending_outbound.capacity() > EMPTY_QUEUE_SHRINK_THRESHOLD {
+            self.pending_outbound.shrink_to_fit();
         }
 
-        if self.inbound.is_empty() && self.keep_alive.is_yes() {
+        // if self.inbound_sender.is_empty() && self.keep_alive.is_yes() { TODO: Fix keep-alive tracking first?
```

Comment:
Blocked by https://github.com/libp2p/rust-libp2p/pull/4029.

Reply:
This isn't necessarily blocked actually. Perhaps we can remove some items from this dependency chain here but finding a different solution.